In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [ ]:
data = pd.read_json('anime.json', encoding='utf-8')

In [ ]:
data.head()

name  \
0                            Dark Gathering   
1                          Trò Chơi Nói Dối   
2                           Horimiya: Piece   
3                                    Temple   
4  Tôi Có Thực Sự Là Người Mạnh Nhất Không?   

                                         description  voting duration  \
0  \nKeitarou, chàng trai có khả năng thu hút ma ...      95    01/25   
1  \nĐảo học viện, nơi các học sinh mở ra những t...      94    02/12   
2  \nChuyển thể những phần ở manga bị cắt trong T...      91    02/13   
3  \nĐúng vậy. Tôi sẽ trở thành một nhà sư! Không...      97    01/12   
4  \nMột thanh niên bị bắt nạt được tái sinh ở th...      89    02/12   

               view                                    status  \
0   36,088 Lượt Xem   Phim đang chiếu/Cập Nhật Tập 01 VietSub   
1  173,964 Lượt Xem   Phim đang chiếu/Cập Nhật Tập 02 VietSub   
2  584,019 Lượt Xem   Phim đang chiếu/Cập Nhật Tập 02 VietSub   
3   86,385 Lượt Xem   Phim đang chiếu/Cập Nhật Tập 01 VietSub   
4  555,460 Lượt Xem   Phim đang chiếu/Cập Nhật Tập 02 VietSub   

                                               genre          director  \
0          [Anime bộ, Shounen, Supernatural, Horror]   Hiroshi Ikehata   
1  [Anime bộ, Sci-Fi, School, Romance, Psychologi...       Oono Satoru   
2  [Anime bộ, Shounen, School, Romance, Slice of ...  Ishihama Masashi   
3  [Anime bộ, Romance, Slice of Life, Harem, Ecch...      Koga Kazuomi   
4  [Anime bộ, Shounen, Romance, Fantasy, Comedy, ...     Naoya Takashi   

    country follower quality                         rating  \
0  Nhật Bản       34      HD   R - 17+ (bạo lực và tục tĩu)   
1  Nhật Bản      222      HD  PG-13 - Teens 13 tuổi trở lên   
2  Nhật Bản      497      HD  PG-13 - Teens 13 tuổi trở lên   
3  Nhật Bản      139      HD  PG-13 - Teens 13 tuổi trở lên   
4  Nhật Bản      643      HD  PG-13 - Teens 13 tuổi trở lên   

                  studio         season  
0                    OLM  Mùa Hạ - 2023  
1              GEEK TOYS  Mùa Hạ - 2023  
2            CloverWorks  Mùa Hạ - 2023  
3                 Gekkou  Mùa Hạ - 2023  
4   Staple Entertainment  Mùa Hạ - 2023

In [ ]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4862 entries, 0 to 4861
Data columns (total 14 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   name         4862 non-null   object
 1   description  4862 non-null   object
 2   voting       4862 non-null   int64 
 3   duration     4862 non-null   object
 4   view         4862 non-null   object
 5   status       4862 non-null   object
 6   genre        4862 non-null   object
 7   director     2585 non-null   object
 8   country      4850 non-null   object
 9   follower     4862 non-null   object
 10  quality      4862 non-null   object
 11  rating       4862 non-null   object
 12  studio       4556 non-null   object
 13  season       4556 non-null   object
dtypes: int64(1), object(13)
memory usage: 531.9+ KB


In [ ]:
data.isnull().sum()

name              0
description       0
voting            0
duration          0
view              0
status            0
genre             0
director       2277
country          12
follower          0
quality           0
rating            0
studio          306
season          306
dtype: int64

**Nhận xét tổng quan**<br>
- Dữ liệu gồm 4862 bản ghi với 14 cột
- Có 4 trường dữ liệu chứa giá trị null là **director, country, studio và season**
- Có 1 trường dữ liệu thuộc dạng int, 13 trường còn lại thuộc dạng object

# Data cleaning

Ta sẽ lần lượt xử lý từng trường dữ liệu có trong dataframe. <br>

<b> Xem xét 1 bản ghi mẫu từ file json </b> <br>

<i>
    {"name": "Fantasista Doll",<br> "description": "\nGái và phép thuật.\nCâu chuyện kể về những em búp bê moe cực độ, sử dụng lá bài để chiến đấu.\n",<br> "voting": "77",<br>"duration": "12/12", <br>"view": "19,788 Lượt Xem",<br> "status": " Full",<br> "genre": ["Anime bộ", "Sci-Fi", "Fantasy", "Magic"], <br>"director": "Taniguchi Gorou",<br> "country": "Nhật Bản", <br>"follower": " 16",<br> "quality": "HD",<br> "rating": "PG-13 - Teens 13 tuổi trở lên",<br> "studio": " Hoods Entertainment, Production IMS",<br> "season": "Mùa Hạ - 2013"},
</i>

---


Có thể thấy 1 số vấn đề như thừa kí tự/khoảng trống. Ta sẽ xử lý chúng trước tiên. Sau đó xem xét đến <br> xử lý các giá trị null -> kiểu dữ liệu -> khác. <br>


In [ ]:
data['description'] = data['description'].str.replace('\n','')
data['view'] = data['view'].str.replace(' Lượt Xem','')
data['view'] = data['view'].str.replace(',','')
data['status'] = data['status'].str.strip()
data['follower'] = data['follower'].str.strip()
data['studio'] = data['studio'].str.strip()

In [ ]:
data.head()

name  \
0                            Dark Gathering   
1                          Trò Chơi Nói Dối   
2                           Horimiya: Piece   
3                                    Temple   
4  Tôi Có Thực Sự Là Người Mạnh Nhất Không?   

                                         description  voting duration    view  \
0  Keitarou, chàng trai có khả năng thu hút ma qu...      95    01/25   36088   
1  Đảo học viện, nơi các học sinh mở ra những trậ...      94    02/12  173964   
2  Chuyển thể những phần ở manga bị cắt trong TV ...      91    02/13  584019   
3  Đúng vậy. Tôi sẽ trở thành một nhà sư! Không b...      97    01/12   86385   
4  Một thanh niên bị bắt nạt được tái sinh ở thế ...      89    02/12  555460   

                                    status  \
0  Phim đang chiếu/Cập Nhật Tập 01 VietSub   
1  Phim đang chiếu/Cập Nhật Tập 02 VietSub   
2  Phim đang chiếu/Cập Nhật Tập 02 VietSub   
3  Phim đang chiếu/Cập Nhật Tập 01 VietSub   
4  Phim đang chiếu/Cập Nhật Tập 02 VietSub   

                                               genre          director  \
0          [Anime bộ, Shounen, Supernatural, Horror]   Hiroshi Ikehata   
1  [Anime bộ, Sci-Fi, School, Romance, Psychologi...       Oono Satoru   
2  [Anime bộ, Shounen, School, Romance, Slice of ...  Ishihama Masashi   
3  [Anime bộ, Romance, Slice of Life, Harem, Ecch...      Koga Kazuomi   
4  [Anime bộ, Shounen, Romance, Fantasy, Comedy, ...     Naoya Takashi   

    country follower quality                         rating  \
0  Nhật Bản       34      HD   R - 17+ (bạo lực và tục tĩu)   
1  Nhật Bản      222      HD  PG-13 - Teens 13 tuổi trở lên   
2  Nhật Bản      497      HD  PG-13 - Teens 13 tuổi trở lên   
3  Nhật Bản      139      HD  PG-13 - Teens 13 tuổi trở lên   
4  Nhật Bản      643      HD  PG-13 - Teens 13 tuổi trở lên   

                 studio         season  
0                   OLM  Mùa Hạ - 2023  
1             GEEK TOYS  Mùa Hạ - 2023  
2           CloverWorks  Mùa Hạ - 2023  
3                Gekkou  Mùa Hạ - 2023  
4  Staple Entertainment  Mùa Hạ - 2023

###Xử lý các giá trị null

####1. Director

In [ ]:
pd.set_option('display.max_rows', None)
data['director'].value_counts(dropna=False)

Trường dữ liệu này thiếu khá nhiều dữ liệu nên khó có thể xem xét cụ thể từng bản ghi. Mặt khác, **director** là cố định cho mỗi bộ phim, không thể suy ra từ các trường dữ liệu khác. Do đó, ta sẽ lựa chọn giá trị mang ý nghĩa chưa rõ để điền vào các giá trị null. <br>

Ngoài ra, có một số giá trị nên đưa về đồng nhất như ***Đang cập nhật, Đang Cập Nhập, Đang cập nhập, Chưa Cập Nhập,...***.

In [ ]:
def normalize_text(x):
    texts = ['Chưa Cập Nhập', 'Đang Cập Nhập', 'Đang cập nhập', 'Chưa rõ', 'Đang Cập Nhật', 'None found, add some']
    for i in range(len(texts)):
        if texts[i] in x:
            x = 'Đang cập nhật'
    return x

In [ ]:
data['director'] = data['director'].fillna('Đang cập nhật')
data['director'] = data['director'].apply(lambda x: normalize_text(x))

####2. Country

In [ ]:
data['country'].value_counts(dropna=False)

Nhật Bản      4523
Trung Quốc     292
Mỹ              29
None            12
Hàn Quốc         5
Việt Nam         1
Name: country, dtype: int64

In [ ]:
data[data['country'].isnull()]

name  \
602                Dogs: Stray Dogs Howling In The Dark   
1683                          Trò Chơi Sinh Tử Specials   
2466                             GOSEI SENTAI DAIRANGER   
2499           Kamen Rider Build Original Drama - ROGUE   
2538                              Eden 2021 (TV series)   
3221                         Lực Bạt Sơn Hà Hề Tử Đường   
3357                                   Vũ Động Càn Khôn   
3705                          Devil May Cry 4 the movie   
4121                                BoBoiBoy: The Movie   
4394                                             Sintel   
4624          Thám Tử Lừng Danh Conan Các Phần Đặc Biệt   
4625  Detective Conan: Kudo Shinichi e no Chousenjou...   

                                            description  voting  \
602   Đó là khoảng thời gian trong tương lai, một th...      55   
1683  Mấy chuyện trên trời dưới đất trong No Game No...      96   
2466  Super Sentai phần 17 sản xuất năm 1993. 6000 n...      10   
2499  Một ngoại truyện nhỏ về Himuro Gentoku gồm 3 p...      73   
2538  Eden Phần 1 - Eden Season 1 (2021) lấy bối cản...      82   
3221  Đông Thánh Tử Đường, dùng võ lập quốc, trong b...      90   
3357  Một đường tu luyện, trộm âm dương, đoạt vận ma...      93   
3705                            Movie của devil may cry      55   
4121  Một nhóm siêu anh hùng nhí có năng lượng và sứ...     100   
4394  Bộ phim mở màn với cảnh một cô gái có tên gọi ...       0   
4624  Các phần đặt biệt của bộ phim hoạt hình conan ...      95   
4625                    Live action của Detective Conan      13   

             duration     view                                   status  \
602               4/4    13195                                     FULL   
1683              5/6   198089   Phim đang chiếu/Cập Nhật Tập 5 VietSub   
2466            50/50    49961                                     Full   
2499            03/??    95531   Phim đang chiếu/Cập Nhật Tập 3 VietSub   
2538            04/04    10798                                 Tập Full   
3221            20/20    99728                                     Full   
3357            36/??  1128438  Phim đang chiếu/Cập Nhật Tập 36 VietSub   
3705              3/3    15804                       Trọn bộ SD VietSub   
4121             Full    25047                       Trọn bộ HD VietSub   
4394  14 phút 48 giây     9828                       Trọn bộ HD VietSub   
4624            08/08   339545                       Trọn bộ HD VietSub   
4625            01/01    37158                       Trọn bộ HD VietSub   

                                  genre           director country follower  \
602                          [Anime bộ]      Đang cập nhật    None        2   
1683   [Anime bộ, Game, Fantasy, Ecchi]      Đang cập nhật    None       94   
2466                         [Anime bộ]      Đang cập nhật    None       40   
2499                 [Anime bộ, Action]      Đang cập nhật    None       26   
2538                         [Anime bộ]      Yasuhiro Irie    None        6   
3221                         [Anime bộ]      Đang cập nhật    None      169   
3357        [Anime bộ, Fantasy, Action]      Đang cập nhật    None    1,458   
3705                         [Anime lẻ]     Hideaki Itsuno    None        1   
4121  [Anime lẻ, Ecchi, Demons, Action]        Nizam Razak    None       20   
4394                [Anime lẻ, Fantasy]      Đang cập nhật    None        1   
4624                         [Anime lẻ]      Đang cập nhật    None      149   
4625                         [Anime lẻ]  Tekeshi Shirakawa    None       16   

     quality                         rating                            studio  \
602       HD                            N/A                          Madhouse   
1683      HD  PG-13 - Teens 13 tuổi trở lên                          Madhouse   
2466      SD                            N/A                              None   
2499      HD  R+ - Dành cho 16 tuổi trở lên                              None   
25

Đặc tính country chỉ thiếu 12 bản ghi nên có thể xem xét từng giá trị để lựa chọn phương án phù hợp. <br>
Country chủ yếu là đất nước nơi anime được sản xuất. Do đó từ **studio** có thể suy ra đặc tính **country** còn thiếu. Madhouse, Qubic Pictures, TMS Entertainment là studio của Nhật Bản, Fengyun Animation là studio ở Trung Quốc và Animonsta Studios là của Malaysia. <br>

In [ ]:
data.loc[data['studio'] == 'Madhouse', 'country'] = 'Nhật Bản'
data.loc[data['studio'] == 'Qubic Pictures, CGCG Studio Inc.', 'country'] = 'Nhật Bản'
data.loc[data['studio'] == 'TMS Entertainment', 'country'] = 'Nhật Bản'
data.loc[data['studio'] == 'Aurora,Fengyun Animation', 'country'] = 'Trung Quốc'
data.loc[data['studio'] == 'Animonsta Studios', 'country'] = 'Malaysia'

Các bản ghi còn lại, dựa theo tên, nội dung tìm kiếm, Vũ Động Càn Khôn là phim Trung Quốc, Sintel thuộc Hà Lan, còn lại đều thuộc Nhật Bản

In [ ]:
data.loc[data['name'] == 'Vũ Động Càn Khôn', 'country'] = 'Trung Quốc'
data.loc[data['name'] == 'Sintel', 'country'] = 'Hà Lan'
data['country'] = data['country'].fillna('Nhật Bản')

####3. Studio

In [ ]:
data['studio'].value_counts(dropna=False)

Tương tự director, studio cũng không thể suy luận từ dữ liệu khác.

In [ ]:
data['studio'] = data['studio'].fillna('Đang cập nhật')
data['studio'] = data['studio'].apply(lambda x: normalize_text(x))

####4. Season

In [ ]:
data['season'].value_counts(dropna=False)


Có một số giá trị khác biệt so với phần còn lại nên ta sẽ thay thế/loại bỏ.

In [ ]:
data.loc[data['season'] == 'Mùa Hạ - 07 - 2021', 'season'] = 'Mùa Hạ - 2021'
data.loc[data['season'] == 'Mùa Hạ - 2010-2011', 'season'] = 'Mùa Hạ - 2010'
data = data.drop(data[data['season'] == 'Mùa Xuân - 2099'].index)

Có thể tách trường dữ liệu này thành 2 trường riêng là year và season.

In [ ]:
data['season'] = data['season'].apply(lambda x: str(x).split(' - ') if x is not None else x)
data['year'] = data['season'].apply(lambda x: x[1] if x is not None else x)
data['season'] = data['season'].apply(lambda x: x[0] if x is not None else x)

In [ ]:
data['year'].head()

0    2023
1    2023
2    2023
3    2023
4    2023
Name: year, dtype: object

In [ ]:
data['season'].head()

0    Mùa Hạ
1    Mùa Hạ
2    Mùa Hạ
3    Mùa Hạ
4    Mùa Hạ
Name: season, dtype: object

In [ ]:
data['year'] = data['year'].fillna('')
data['season'] = data['season'].fillna('')

In [ ]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 4861 entries, 0 to 4861
Data columns (total 15 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   name         4861 non-null   object
 1   description  4861 non-null   object
 2   voting       4861 non-null   int64 
 3   duration     4861 non-null   object
 4   view         4861 non-null   object
 5   status       4861 non-null   object
 6   genre        4861 non-null   object
 7   director     4861 non-null   object
 8   country      4861 non-null   object
 9   follower     4861 non-null   object
 10  quality      4861 non-null   object
 11  rating       4861 non-null   object
 12  studio       4861 non-null   object
 13  season       4861 non-null   object
 14  year         4861 non-null   object
dtypes: int64(1), object(14)
memory usage: 607.6+ KB


###Xử lý kiểu dữ liệu

Các trường dữ liệu ***view, follower*** nên là kiểu **int**

In [ ]:
data['follower'] = data['follower'].str.replace(',','')

In [ ]:
data['follower'] = data['follower'].astype(int)
data['view'] = data['view'].astype(int)

**voting** là tỉ lệ người theo dõi bình chọn cho anime, do đó trường dữ liệu này nên đưa về tỉ lệ %

In [ ]:
data['voting'] = data['voting']/100

In [ ]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 4861 entries, 0 to 4861
Data columns (total 15 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   name         4861 non-null   object 
 1   description  4861 non-null   object 
 2   voting       4861 non-null   float64
 3   duration     4861 non-null   object 
 4   view         4861 non-null   int64  
 5   status       4861 non-null   object 
 6   genre        4861 non-null   object 
 7   director     4861 non-null   object 
 8   country      4861 non-null   object 
 9   follower     4861 non-null   int64  
 10  quality      4861 non-null   object 
 11  rating       4861 non-null   object 
 12  studio       4861 non-null   object 
 13  season       4861 non-null   object 
 14  year         4861 non-null   object 
dtypes: float64(1), int64(2), object(12)
memory usage: 607.6+ KB


In [ ]:
data.head()

name  \
0                            Dark Gathering   
1                          Trò Chơi Nói Dối   
2                           Horimiya: Piece   
3                                    Temple   
4  Tôi Có Thực Sự Là Người Mạnh Nhất Không?   

                                         description  voting duration    view  \
0  Keitarou, chàng trai có khả năng thu hút ma qu...    0.95    01/25   36088   
1  Đảo học viện, nơi các học sinh mở ra những trậ...    0.94    02/12  173964   
2  Chuyển thể những phần ở manga bị cắt trong TV ...    0.91    02/13  584019   
3  Đúng vậy. Tôi sẽ trở thành một nhà sư! Không b...    0.97    01/12   86385   
4  Một thanh niên bị bắt nạt được tái sinh ở thế ...    0.89    02/12  555460   

                                    status  \
0  Phim đang chiếu/Cập Nhật Tập 01 VietSub   
1  Phim đang chiếu/Cập Nhật Tập 02 VietSub   
2  Phim đang chiếu/Cập Nhật Tập 02 VietSub   
3  Phim đang chiếu/Cập Nhật Tập 01 VietSub   
4  Phim đang chiếu/Cập Nhật Tập 02 VietSub   

                                               genre          director  \
0          [Anime bộ, Shounen, Supernatural, Horror]   Hiroshi Ikehata   
1  [Anime bộ, Sci-Fi, School, Romance, Psychologi...       Oono Satoru   
2  [Anime bộ, Shounen, School, Romance, Slice of ...  Ishihama Masashi   
3  [Anime bộ, Romance, Slice of Life, Harem, Ecch...      Koga Kazuomi   
4  [Anime bộ, Shounen, Romance, Fantasy, Comedy, ...     Naoya Takashi   

    country  follower quality                         rating  \
0  Nhật Bản        34      HD   R - 17+ (bạo lực và tục tĩu)   
1  Nhật Bản       222      HD  PG-13 - Teens 13 tuổi trở lên   
2  Nhật Bản       497      HD  PG-13 - Teens 13 tuổi trở lên   
3  Nhật Bản       139      HD  PG-13 - Teens 13 tuổi trở lên   
4  Nhật Bản       643      HD  PG-13 - Teens 13 tuổi trở lên   

                 studio  season  year  
0                   OLM  Mùa Hạ  2023  
1             GEEK TOYS  Mùa Hạ  2023  
2           CloverWorks  Mùa Hạ  2023  
3                Gekkou  Mùa Hạ  2023  
4  Staple Entertainment  Mùa Hạ  2023

###Xử lý một số trường dữ liệu khác

####1. Rating

In [ ]:
data['rating'].value_counts(dropna=False)

PG-13 - Teens 13 tuổi trở lên    2851
R - 17+ (bạo lực và tục tĩu)      564
None                              515
R+ - Dành cho 16 tuổi trở lên     365
G - Mọi lứa tuổi                  220
PG - Trẻ em                       192
N/A                               133
PG-13 - Teens 13 or older           6
Đang Cập Nhập                       3
13+ - Teens 13 or older             3
R+ 17                               2
undefined                           1
R + - Ảnh khỏa thân nhẹ             1
Rx - Hentai                         1
PG-13                               1
Nona                                1
PG - Children                       1
Đang cập nhập                       1
Name: rating, dtype: int64

Ta sẽ đồng nhất một số giá trị tương đương với nhau

In [ ]:
def normalize_rating(x):
    if x == 'PG-13' or x == 'PG-13 - Teens 13 or older':
        x = 'PG-13 - Teens 13 tuổi trở lên'
        return x

    if x == '13+ - Teens 13 or older':
        x = '13+ - Teens 13 tuổi trở lên'
        return x

    if x == 'PG - Children':
        x = 'PG - Trẻ em'
        return x

    if x == 'Đang cập nhập' or x == 'Đang Cập Nhập' or x == 'Nona' or x == 'N/A' or x == 'None' or x == 'undefined':
        x = 'Đang cập nhật'
        return x

    return x

In [ ]:
data['rating'] = data['rating'].apply(lambda x: normalize_rating(x))

In [ ]:
data['rating'].value_counts(dropna=False)

PG-13 - Teens 13 tuổi trở lên    2857
Đang cập nhật                     654
R - 17+ (bạo lực và tục tĩu)      564
R+ - Dành cho 16 tuổi trở lên     365
G - Mọi lứa tuổi                  220
PG - Trẻ em                       193
13+ - Teens 13 tuổi trở lên         3
R+ 17                               2
Rx - Hentai                         1
R + - Ảnh khỏa thân nhẹ             1
Name: rating, dtype: int64

####2. Quality

In [ ]:
data['quality'].value_counts(dropna=False)

HD                4163
BD                 279
SD                 230
FHD                122
BD/Bluray           33
HQ                  23
Full HD              2
hd                   1
26/10/2018           1
None                 1
Bluray Disc          1
rHD                  1
CAM HD               1
BD Vietsub           1
OVA BD Vietsub       1
480P                 1
Name: quality, dtype: int64

In [ ]:
data['quality'] = data['quality'].replace('hd', 'HD')
data = data.drop(data[data['quality'] == '26/10/2018'].index)

In [ ]:
data['quality'].value_counts(dropna=False)

HD                4164
BD                 279
SD                 230
FHD                122
BD/Bluray           33
HQ                  23
Full HD              2
None                 1
Bluray Disc          1
rHD                  1
CAM HD               1
BD Vietsub           1
OVA BD Vietsub       1
480P                 1
Name: quality, dtype: int64

####3. Genre

Dựa trên **genre**, có thể tách thêm thuộc tính thể loại bao gồm phim bộ/lẻ

In [ ]:
data['type'] = data['genre'].apply(lambda x: x.pop(0))

In [ ]:
data.head()

name  \
0                            Dark Gathering   
1                          Trò Chơi Nói Dối   
2                           Horimiya: Piece   
3                                    Temple   
4  Tôi Có Thực Sự Là Người Mạnh Nhất Không?   

                                         description  voting duration    view  \
0  Keitarou, chàng trai có khả năng thu hút ma qu...    0.95    01/25   36088   
1  Đảo học viện, nơi các học sinh mở ra những trậ...    0.94    02/12  173964   
2  Chuyển thể những phần ở manga bị cắt trong TV ...    0.91    02/13  584019   
3  Đúng vậy. Tôi sẽ trở thành một nhà sư! Không b...    0.97    01/12   86385   
4  Một thanh niên bị bắt nạt được tái sinh ở thế ...    0.89    02/12  555460   

                                    status  \
0  Phim đang chiếu/Cập Nhật Tập 01 VietSub   
1  Phim đang chiếu/Cập Nhật Tập 02 VietSub   
2  Phim đang chiếu/Cập Nhật Tập 02 VietSub   
3  Phim đang chiếu/Cập Nhật Tập 01 VietSub   
4  Phim đang chiếu/Cập Nhật Tập 02 VietSub   

                                               genre          director  \
0                    [Shounen, Supernatural, Horror]   Hiroshi Ikehata   
1  [Sci-Fi, School, Romance, Psychological, Game,...       Oono Satoru   
2  [Shounen, School, Romance, Slice of Life, Comedy]  Ishihama Masashi   
3     [Romance, Slice of Life, Harem, Ecchi, Comedy]      Koga Kazuomi   
4     [Shounen, Romance, Fantasy, Comedy, Adventure]     Naoya Takashi   

    country  follower quality                         rating  \
0  Nhật Bản        34      HD   R - 17+ (bạo lực và tục tĩu)   
1  Nhật Bản       222      HD  PG-13 - Teens 13 tuổi trở lên   
2  Nhật Bản       497      HD  PG-13 - Teens 13 tuổi trở lên   
3  Nhật Bản       139      HD  PG-13 - Teens 13 tuổi trở lên   
4  Nhật Bản       643      HD  PG-13 - Teens 13 tuổi trở lên   

                 studio  season  year      type  
0                   OLM  Mùa Hạ  2023  Anime bộ  
1             GEEK TOYS  Mùa Hạ  2023  Anime bộ  
2           CloverWorks  Mùa Hạ  2023  Anime bộ  
3                Gekkou  Mùa Hạ  2023  Anime bộ  
4  Staple Entertainment  Mùa Hạ  2023  Anime bộ

####4. Status

Với dữ liệu **status** hiện tại, có thể thấy đối với các phim chưa ra full, trạng thái *Phim đang chiếu* đã đủ để khai thác, số tập đang ra chưa quá quan trọng nên có thể bỏ qua

In [ ]:
data[data['type'] == 'Anime lẻ']['status'].value_counts(dropna=False)

Trọn bộ HD VietSub                  955
Trọn bộ SD VietSub                   84
Trọn bộ BD VietSub                   65
Trọn bộ FHD VietSub                  48
Trọn bộ BD/Bluray VietSub             7
Trọn bộ VietSub                       6
Trọn bộ HQ VietSub                    5
Trọn bộ Full HD VietSub               2
Trọn bộ CAM HD VietSub                1
Trailer                               1
Trọn bộ BD Vietsub VietSub            1
Trọn bộ OVA BD Vietsub VietSub        1
Trọn bộ HD VietSub + Thuyết minh      1
Trọn bộ 480P VietSub                  1
Trọn bộ HD Thuyết minh                1
Name: status, dtype: int64

In [ ]:
data[data['type'] == 'Anime bộ']['status'].value_counts(dropna=False)

In [ ]:
def normalize_status(x):
    texts = ['Trọn bộ', 'Full', 'FULL', 'END']

    for i in range(len(texts)):
        if texts[i] in x:
            x = 'Trọn bộ'
            return x

    x = 'Phim đang chiếu'
    return x

In [ ]:
data['status'] = data['status'].apply(lambda x: normalize_status(x))

In [ ]:
data['status'].value_counts()

Trọn bộ            4336
Phim đang chiếu     524
Name: status, dtype: int64

####5. Duration

Đối với trường dữ liệu **duration**, có nhiều vấn đề cần phải xử lý. Đầu tiên, định dạng cho phim lẻ đang bao gồm cả giờ, phút, giây nên cần đưa về 1 đơn vị phù hợp để dễ dàng phân tích (giây). Ngoài ra, với phim gồm nhiều tập đã ra full, cần đưa về đơn vị số tập.

In [ ]:
data['duration'] = data['duration'].str.lower()

In [ ]:
data[data['type'] == 'Anime lẻ']['duration'].value_counts(dropna=False)

In [ ]:
data[data['type'] == 'Anime bộ']['duration'].value_counts(dropna=False)

In [ ]:
def normalize_duration(x):
    if x == '1h 35p':
        x = '1 giờ 35 phút'

    if x == '1 hr 59 min':
        x = '1 giờ 59 phút'

    if 'tiếng' in x:
        x = x.replace('tiếng', 'giờ')

    return x

In [ ]:
data['duration'] = data['duration'].apply(lambda x: normalize_duration(x))

In [ ]:
def convert_to_minute(x):
    time = 0
    value = ''
    if '/' not in x:
        if 'giờ' in x or 'phút' in x or 'giây' in x:
            x = x.split()

            if 'giờ' in x:
                i = x.index('giờ')
                time += int(x[i-1])*60

            if 'phút' in x:
                i = x.index('phút')
                time += int(x[i-1])

            if 'giây' in x:
                i = x.index('giây')
                time += round(int(x[i-1])*1/60, 2)

            value = str(time) + ' phút'
        else:
            value = x
    else:
        value = x

    return value

In [ ]:
data['duration'] = data['duration'].apply(lambda x: convert_to_minute(x))

In [ ]:
# check giá trị khác biệt
for value in data['duration']:
    if ('?' not in value) & ('phút' not in value) & (' ' in value) or ('ova' in value):
        print(value)
        print('\n')

In [ ]:
data['duration'] = data['duration'].replace('25/25 + ova', '26 tập')
data['duration'] = data['duration'].replace('14 + ova/14', '15 tập')
data['duration'] = data['duration'].replace('sp 02/10', '11 tập')
data['duration'] = data['duration'].replace('12/12 + ova', '13 tập')
data['duration'] = data['duration'].replace('4/4 + ova', '5 tập')
data['duration'] = data['duration'].replace('26/26 + 2 sp + 7 ngoại truyện', '35 tập')
data['duration'] = data['duration'].replace('11/11 + sp', '13 tập')
data['duration'] = data['duration'].replace('08 bd/70', '70 tập')
data['duration'] = data['duration'].replace('12/12 + 2 ova', '14 tập')
data['duration'] = data['duration'].replace('12 + 2 ova/12', '13 tập')
data['duration'] = data['duration'].replace('12/12 + 04 sp', '16 tập')
data['duration'] = data['duration'].replace('chưa biết', 'chưa rõ')
data['duration'] = data['duration'].replace('s1 + 13 s2', 'chưa rõ')
data['duration'] = data['duration'].replace('s1 + ova', 'chưa rõ')
data['duration'] = data['duration'].replace('11/11+ova', '12 tập')
data['duration'] = data['duration'].replace('24/24+ova', '25 tập')

In [ ]:
def convert_to_episode(x):
    value = ''
    if ('/' in x):
        if '?' in x:
            value = 'chưa rõ'
        else:
            x = x.split('/')
            value = x[1] + ' tập'
    else:
        value = x
    return value

In [ ]:
data['duration'] = data['duration'].apply(lambda x: convert_to_episode(x))

In [ ]:
for x in data['duration']:
    if ('tập' not in x) & ('phút' not in x) & ('full' not in x) & (x != 'chưa rõ'):
        print(x)
        print('\n')

12


26


26


13


13


01


12




In [ ]:
data['duration'] = data['duration'].replace('12', '12 tập')
data['duration'] = data['duration'].replace('26', '26 tập')
data['duration'] = data['duration'].replace('13', '13 tập')
data['duration'] = data['duration'].replace('01', '1 tập')

In [ ]:
data.head()

name  \
0                            Dark Gathering   
1                          Trò Chơi Nói Dối   
2                           Horimiya: Piece   
3                                    Temple   
4  Tôi Có Thực Sự Là Người Mạnh Nhất Không?   

                                         description  voting duration    view  \
0  Keitarou, chàng trai có khả năng thu hút ma qu...    0.95   25 tập   36088   
1  Đảo học viện, nơi các học sinh mở ra những trậ...    0.94   12 tập  173964   
2  Chuyển thể những phần ở manga bị cắt trong TV ...    0.91   13 tập  584019   
3  Đúng vậy. Tôi sẽ trở thành một nhà sư! Không b...    0.97   12 tập   86385   
4  Một thanh niên bị bắt nạt được tái sinh ở thế ...    0.89   12 tập  555460   

            status                                              genre  \
0  Phim đang chiếu                    [Shounen, Supernatural, Horror]   
1  Phim đang chiếu  [Sci-Fi, School, Romance, Psychological, Game,...   
2  Phim đang chiếu  [Shounen, School, Romance, Slice of Life, Comedy]   
3  Phim đang chiếu     [Romance, Slice of Life, Harem, Ecchi, Comedy]   
4  Phim đang chiếu     [Shounen, Romance, Fantasy, Comedy, Adventure]   

           director   country  follower quality  \
0   Hiroshi Ikehata  Nhật Bản        34      HD   
1       Oono Satoru  Nhật Bản       222      HD   
2  Ishihama Masashi  Nhật Bản       497      HD   
3      Koga Kazuomi  Nhật Bản       139      HD   
4     Naoya Takashi  Nhật Bản       643      HD   

                          rating                studio  season  year      type  
0   R - 17+ (bạo lực và tục tĩu)                   OLM  Mùa Hạ  2023  Anime bộ  
1  PG-13 - Teens 13 tuổi trở lên             GEEK TOYS  Mùa Hạ  2023  Anime bộ  
2  PG-13 - Teens 13 tuổi trở lên           CloverWorks  Mùa Hạ  2023  Anime bộ  
3  PG-13 - Teens 13 tuổi trở lên                Gekkou  Mùa Hạ  2023  Anime bộ  
4  PG-13 - Teens 13 tuổi trở lên  Staple Entertainment  Mùa Hạ  2023  Anime bộ

In [ ]:
data.to_csv('anime_cleaned.csv', encoding = 'utf-8-sig')